### DISCUSSION TITLE
Unable to add users to invite-only batches

### DISCUSSION THREAD

**Discussion by Xitija** [Sunbird-Lern/Community#141] [Issues] [answered]
When I am creating a batch with type as invite-only and submitting the form I can see the API for adding users to be failing
API - `/learner/course/v1/batch/user/add/01383834429512908820`
Response - `{"message":"no route and no API found with those values"}`

Can anyone please tell me what is going wrong?
 
cc : @coolbung @komalm @Renuka-S @vinod303 @paritshivani @reshmi-nair

**Comment by Xitija** [comment, marked as answer]
I changed the endpoint from local to dev then I am getting the following response
`
{
    "id": "api.error",
    "ver": "1.0",
    "ts": "2023-07-17 11:11:18:963+0000",
    "params": {
        "resmsgid": "a7888030-2492-11ee-8f03-ff5607f12343",
        "msgid": null,
        "status": "failed",
        "err": "FORBIDDEN_ERROR",
        "errmsg": "Forbidden: API WHITELIST Access is denied"
    },
    "responseCode": "FORBIDDEN",
    "result": {}
}
`
Can anyone please guide me here?

**Reply by AmiableAnil** [reply]
@Xitija Can you please share the curl, which you are using.

**Reply by Xitija** [reply]
Sure, here it is. 
`curl --location --request POST 'https://devnulp.niua.org/learner/course/v1/batch/user/add/01384114526933811237' \
--header 'Accept: application/json' \
--header 'Accept-Language: en-GB,en;q=0.9' \
--header 'Connection: keep-alive' \
--header 'Content-Type: application/json' \
--header 'Cookie: connect.sid=s%3AZpcpVAcKJ8nbPjkwrjEDrs-0dMT3JTNq.dBGWsXVnbo35JssvCe0ZAkY9Dw3hEz7YM%2FRaz2fZXCE; connect.sid=s%3AtOJ0U2xjJ-pEppSdLgYBMYuVocTfqTsc.3n8ejag3LF6EbWSqfLZbvrHBbhJxYcYji0foDftg7ps' \
--header 'Origin: http://localhost:3000' \
--header 'Referer: http://localhost:3000/learn/course/do_1138388300769525761143/update/batch/01384114526933811237?enrollmentType=invite-only' \
--header 'Sec-Fetch-Dest: empty' \
--header 'Sec-Fetch-Mode: cors' \
--header 'Sec-Fetch-Site: same-origin' \
--header 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36' \
--header 'X-App-Id: local.kshitija.portal' \
--header 'X-App-Version: 4.10.0' \
--header 'X-Channel-Id: 0130701891041689600' \
--header 'X-Device-ID: 4e22d793cffea9fb57f2a2e392877115' \
--header 'X-Org-code: 0130701891041689600' \
--header 'X-Request-ID: 6d7dea99-f5f7-328d-de6c-69689b99f0a0' \
--header 'X-Session-ID: ZpcpVAcKJ8nbPjkwrjEDrs-0dMT3JTNq' \
--header 'X-Source: web' \
--header 'X-User-ID: 5d757783-a86a-40cd-a814-1b6a16d37cb6' \
--header 'X-msgid: 6d7dea99-f5f7-328d-de6c-69689b99f0a0' \
--header 'sec-ch-ua: "Not.A/Brand";v="8", "Chromium";v="114", "Google Chrome";v="114"' \
--header 'sec-ch-ua-mobile: ?0' \
--header 'sec-ch-ua-platform: "Linux"' \
--header 'ts: 2023-07-18T11:34:57+05:30' \
--data-raw '{"request":{"userIds":["add88f79-bd54-47ee-8e17-07afe395f1db","010dcb12-1403-41d4-942d-25d414a2db16","05009b07-d77d-4883-a128-be18c1ccb8dc"]}}'
`

**Reply by Xitija** [reply]
@AmiableAnil Sir,
I went through the documentation
https://lern.sunbird.org/learn/product-and-developer-guide/batch-service/api-documentation
But I could not find any detail related to this API, can you please guide me here? 
I need a resolution for this issue.

**Reply by Jayaprakash8887** [reply]
@Xitija What is the Release version you are using? 

Please check if the batch/user/add or similar is part of 'routes' file in course-batch service code. Also, all APIs on boarded will be listed in 
https://github.com/project-sunbird/sunbird-devops/blob/release-5.2.0/kubernetes/ansible/roles/kong-api/defaults/main.yml

**Reply by Jayaprakash8887** [reply]
Most likely, this API might no longer exist in the release version you are using. Might have been deprecated.

**Reply by Xitija** [reply]
@Jayaprakash8887  
Sir, I am on release version 4.10.0
The course-batch.service.ts file contains the following code
```
 addUsersToBatch(request, batchId) {
    const option = {
      url: this.configService.urlConFig.URLS.BATCH.ADD_USERS + '/' + batchId,
      data: {
        request: request
      }
    };
    return this.learnerService.post(option);
  }
```
In URL config it is `"ADD_USERS": "course/v1/batch/user/add"`

**Reply by Jayaprakash8887** [reply]
@Xitija Invite-only batch support seems to be have been deprecated post releas-3.7.0. Ref: https://github.com/Sunbird-Lern/lms-service/pull/335

If you have a forked repository for your project, suggest you to take the code from release-3.7.0 

cc: @rhwarrier @AmiableAnil

**Reply by Xitija** [reply]
As per my knowledge, it was enabled back in the 4.10 version and above.
Can you please help me clarify ?
 @coolbung @komalm @Renuka-S @gandham-santhosh

**Reply by Xitija** [reply]
@Jayaprakash8887 
https://github.com/orgs/Sunbird-Lern/discussions/14
In the above discussion, we got to know invite-only was enabled back in Sunbird with release-4.9.0.

If not in 4.10 then which is the next release in which it is enabled?

Is there a way/process to bring the necessary functions in the release that we are using to get the feature enabled? 

cc: @Renuka-S @komalm @paritshivani

**Reply by Jayaprakash8887** [reply]
@Xitija As per the PR shared in https://github.com/orgs/Sunbird-Lern/discussions/14 , there were new API endpoints introduced for the invite-only batch. These new API endpoints are part of releases above 4.9.0 

![image](https://github.com/Sunbird-Lern/Community/assets/19702246/9627e12d-c3c1-4b36-a818-eccf46b90028)

Guess, you need to use '/v1/course/admin/enroll' instead of the API you are using for adding users.

cc: @AmiableAnil @rhwarrier

**Reply by Xitija** [reply]
@Jayaprakash8887 
Where can I get more information about these endpoints? Is there any document that I can refer?
I thought the enroll apis were related to open courses enrollment, I am not able to figure out which endpoint is used when.

Please guide me here.

**Reply by Jayaprakash8887** [reply]
@Xitija Apologies. I am unable to find necessary documentation for these API endpoints. However, from the code, I could prepare the postman collection. Kindly import the same. 
**Note:** Please note that these APIs are not mentioned in https://github.com/project-sunbird/sunbird-devops/blob/release-5.2.0/kubernetes/ansible/roles/kong-api/defaults/main.yml which means, you will have to make corresponding entries and use "onboardAPIs" Jenkins job while deploying to server.

Please rename file extension to '.json' before importing to Postman

[Invite-only Batch.postman_collection.txt](https://github.com/Sunbird-Lern/Community/files/12212009/Invite-only.Batch.postman_collection.txt)

**Reply by Xitija** [reply]
Thank you.I will have a look at this.

**Reply by Xitija** [reply]
@Jayaprakash8887 

I went through the API Collection
The API `v1/course/admin/enroll` can send only one userId at a time, I want to send multiple userIds for adding them to the batch.

Also what kind of corresponding entries are to be made in the main.yml? Can you please point me to it, I am fairly new to this.

**Reply by Jayaprakash8887** [reply]
@Xitija v1/course/admin/enroll accepts only one user at a time. You will have to make multiple API calls to enroll multiple users.

For main.yml changes, follow/copy the exisiting courseEnrollment related entries and modify accordingly. 

Example:

```
- name: adminUserUnEnrollment
  uris: "{{ course_service_prefix }}/v1/admin/unenroll"
  upstream_url: "{{ lms_service_url }}/v1/course/admin/unenroll"
  strip_uri: true
  plugins:
  - name: jwt
  - name: cors
  - "{{ statsd_pulgin }}"
  - name: acl
    config.whitelist:
    - courseAccess
  - name: rate-limiting
    config.policy: local
    config.hour: "{{ medium_rate_limit_per_hour }}"
    config.limit_by: credential
  - name: request-size-limiting
    config.allowed_payload_size: "{{ small_request_size_limit }}"
  - name: opa-checks
    config.required: true
    config.enabled: true
```

**Reply by Xitija** [reply]
@Jayaprakash8887 

Thank you I will try editing the main.yml file and check the endpoints.

My UI has been developed for adding multiple users to the batch at once.

![Screenshot from 2023-07-31 17-32-13](https://github.com/Sunbird-Lern/Community/assets/65657373/33bc15ba-9562-479c-af40-6455628d5d72)

Which was supported by the endpoint which I was using previously.
The request body looks as follows, is there no way to support multiple ids at once? 
Adding users one by one is really time consuming.
```
{
    "userIds": [
        "34cc4869-aa6a-4cbc-8b40-8d0f2c94a0ac",
        "6bd4faa9-9686-49dc-9918-0cbe64277572",
        "f6f12fe9-9577-45a7-97ba-e004f63480a8"
    ]
}
```

**Reply by Jayaprakash8887** [reply]
@Xitija We can take this enhancement in upcoming releases but we cant go back and edit the code for the release you are using. However, you can do the necessary code changes in your forked repo and contribute back :) 

cc: @AmiableAnil @rhwarrier

**Reply by Xitija** [reply]
@Jayaprakash8887 
I could understand the API calls for user enrol and unenrol but I am not able to figure out what this API is for
```
curl --location --request POST 'https://devnulp.niua.org/v2/user/courses/admin/list?fields=contentType,topic,name,channel&batchDetails=name,endDate,startDate,status,enrollmentType,createdBy' \
--header 'Content-Type: application/json' \
--header 'x-authenticated-user-token: {{keycloak_access_token}}' \
--header 'Authorization: {{kong_api_key}}' \
--data-raw '{
  "request": {
      "userId": "{{user_id}}",
      "courseIds": ["",""]
    }
}'
```
For what and when this API is used, I thought it will return courses user has enrolled, but I am not able to understand why we are passing course Ids array?
